go-controller/pkg/util/kube.go

```diff
@@ -719,51 +733,88 @@ func NoHostSubnet(node *v1.Node) bool {
+// (an endpoint that is terminating shows ready=false). Then, only if no ready endpoints are found,
+// it falls back to serving terminating endpoints.
+// However, if the service has Spec.PublishNotReadyAddresses set, all provided endpoints are always returned.
+// PublishNotReadyAddresses tells endpoint consumers to disregard any indications of ready/not-ready and is generally
+// used together with headless services so that DNS records of all endpoints (ready or not) are always published.
+func GetEligibleEndpoints(endpoints []discovery.Endpoint, service *kapi.Service) []discovery.Endpoint {
 	includeTerminating := service != nil && service.Spec.PublishNotReadyAddresses
-	var shortcut bool
-	for _, endpoint := range endpointSlice.Endpoints {
-		if IsEndpointEligible(endpoint, includeTerminating) {
-			fn(endpoint, &shortcut)
-			if shortcut {
-				// shortcircuit the whole iteration
-				return
+	var eligibleEndpoints []discovery.Endpoint
+
+	for _, endpoint := range endpoints {
+		if includeTerminating || IsEndpointReady(endpoint) {
+			eligibleEndpoints = append(eligibleEndpoints, endpoint)
+		}
+	}
+	serviceStr := ""
+	if service != nil {
+		serviceStr = fmt.Sprintf(" (service %s/%s)", service.Namespace, service.Name)
+	}
+	klog.V(5).Infof("Endpoint selection%s: found %d ready endpoints", serviceStr, len(eligibleEndpoints))
+
+	// Fallback to serving terminating endpoints (ready=false, serving=true, terminating=true) only if none are ready.
+	if len(eligibleEndpoints) == 0 {
+		for _, endpoint := range endpoints {
+			if IsEndpointServing(endpoint) && IsEndpointTerminating(endpoint) {
+				eligibleEndpoints = append(eligibleEndpoints, endpoint)
 			}
 		}
+		klog.V(5).Infof("Endpoint selection%s: fallback to %d serving & terminating endpoints",
+			serviceStr, len(eligibleEndpoints))
+	}
+
+	return eligibleEndpoints
+}
+
+// ForEachEligibleEndpoint first finds eligible endpoints from a provided list of endpoints
+// and then iterates through them to apply the given function fn to each eligible endpoint.
+func ForEachEligibleEndpoint(endpoints []discovery.Endpoint, service *kapi.Service, fn func(ep discovery.Endpoint)) {
+	eligibleEndpoints := GetEligibleEndpoints(endpoints, service)
+	for _, endpoint := range eligibleEndpoints {
+		fn(endpoint)
 	}
 }
 
-// GetEndpointAddresses returns a list of IP addresses of all eligible endpoints in the given endpoint slices.
-func GetEndpointAddressesWithCondition(endpointSlices []*discovery.EndpointSlice, service *kapi.Service, fn func(discovery.Endpoint) bool) sets.Set[string] {
+// GetEndpointAddressesWithPrecondition first applies the (optionally) provided function fn
+// to filter endpoints from the given list of endpoint slices; from the resulting list,
+// it returns the IP addresses of eligible endpoints.
+func GetEndpointAddressesWithPreCondition(endpointSlices []*discovery.EndpointSlice, service *kapi.Service, fn func(discovery.Endpoint) bool) sets.Set[string] {
```

Comment:
May be pedantic but why change from `GetEndpointAddressesWithCondition()`?

Reply:
No, that's that good question :) 
In the previous implementation, the selection of endpoints was easier and had no fallback case: I would just take all ready endpoints + all serving terminating endpoints.  `GetEndpointAddressesWithCondition` first selected endpoints that way and after that it would apply the provided filter function (in our case, to keep only local endpoints). 
With the new selection logic in this PR, the order between the two steps becomes important: I need to first apply the filter function (e.g keep all local endpoints) and only then apply the new selection logic (take all ready endpoints, but if none are ready take serving terminating endpoints). The ordering here changes the result.

Imagine you have three nodes and you need to get the "eligible" endpoints local to node1:
- node1: endpoint1_serving&terminating, endpoint2_serving&terminating
- node2: endpoint3_serving&terminating, endpoint4_ready
- node3: endpoint5_ready, endpoint6_ready

If I first apply the new selection logic and only then keep local endpoints, I would end up with no endpoint:
1. [endpoint selection] all ready endpoints in the cluster: endpoint4, endpoint5, endpoint6
2. [filter function] none, since the ready endpoints are not on node1

If I reverse the steps I get the result I want:
1. [filter function] endpoint1, endpoint2  --> they are local to node1
2. [endpoint selection] endpoint1, endpoint2  --> they are serving & terminating

So I replaced `Condition` with `PreCondition` in the function name to highlight that it gets applied first.

Reply:
I've just added a test case to cover the scenario I described above. 
https://github.com/ovn-org/ovn-kubernetes/pull/4072/files#diff-4c82fea05fd74e2aaa346d3d3caaa27442ed190cf4057e8e1a810982eafdd286R1807